### DISCUSSION TITLE
Installing packages via jpm on NixOS

### DISCUSSION THREAD

**Discussion by nstgc** [janet-lang/janet#1393] [Q&A]
NixOS newbie here!

I installed janet and jpm by declaring them in `environment.systemPackages`, but when I try `sudo jpm install spork` I'm informed that the Nix Store is read-only (same error without `sudo`, too). I know that is the way of NixOS and appreciate it as a good thing, but it does complicate the task at hand.

For those using NixOS, how do you work around this? One thing I'm considering is building from source and running it from `~/.local/bin`. I already have `environment.localBinInPath = true;` set.

**Comment by ianthehenry** [comment]
Can you use `jpm -l install spork` instead of installing packages globally?

If you do want to globally install packages, you might be able to set a different global `JANET_TREE` / `JANET_PATH` in a writeable directory, or modify the derivation to change the default.

(I recall there was some problem a couple years ago building native binaries on NixOS because the jpm derivation was just configured to expect it to be installed to the same root directory as janet... but maybe this is fixed now.)

**Reply by nstgc** [reply]
That does sound like it should work. Thanks!

**Reply by nstgc** [reply]
Actually, this does not seem to work. (Finally getting around to trying again. :sweat_smile: ) I'm told `janet.h` doesn't exist.

What I ended up doing is installing from source into my `~/.local`. Much as I prefer having things managed by my package manager, I've been managing things manually for over a year now on Arch.

**Comment by sogaiu** [comment]
Possibly of interest is [this comment](https://github.com/janet-lang/jpm/issues/68#issuecomment-1836481831).

**Reply by nstgc** [reply]
Ooh, fancy. That's probably beyond me at the moment, but definitely the direction I think I should go in. Thanks!
